client/vscode-lib/src/controller.ts

```diff
@@ -119,75 +104,70 @@ export function createController({
         providers,
     })
 
-    const errorTapObserver: Partial<TapObserver<any>> = {
-        next(): void {
-            errorWaiter.gotError(false)
-        },
-        error(): void {
-            // Show an error notification unless we've recently shown one (to avoid annoying
-            // the user).
-            const shouldNotify = errorWaiter.timeSinceLastError() > 1000 * 60 * 15 /* 15 min */
-            if (shouldNotify) {
-                showErrorNotification(outputChannel)
-            }
-
-            errorWaiter.gotError(true)
-        },
-    }
-    const errorCatcher = <T = any>(error: any): Observable<T[]> => {
+    const errorLog = (error: any) => {
+        console.error(error)
         outputChannel.appendLine(error)
-        return of([])
     }
 
+    // errorReporter contains a lot of logic and state on how we notify and log
+    // errors, as well as state around if we should turn off a feature (see
+    // skipIfImplicitAction)
+    const errorReporter = new ErrorReporterController(
+        createErrorNotifier(outputChannel, extensionId),
+        errorLog
+    )
+    disposables.push(errorReporter)
+
+    // Note: We distingiush between an explicit user action and an implicit
+    // one. Explicit user actions should always run and return errors.
+    // Implicit actions may not run if they are erroring a lot. Currently only
+    // annotations is implicit.
+    //
+    // Note: the client swallows errors, so the observable methods will report
+    // internal errors but the behaviour around skipping is poor.
+
     /**
      * The controller is passed to UI feature providers for them to fetch data.
      */
     const controller: Controller = {
-        observeMeta(params: MetaParams, opts?: ProviderMethodOptions) {
-            if (!errorWaiter.ok()) {
-                return of([])
-            }
-            return client.metaChanges(params, opts).pipe(tap(errorTapObserver), catchError(errorCatcher))
+        observeMeta(params: MetaParams, optsInput?: ProviderMethodOptions) {
+            const { opts, tapAndCatch } = errorReporter.getForObservable(UserAction.Explicit, optsInput)
```

Comment:
wdyt about doing someting like

`errorReporter.handleObservable(UserAction.Explicit, optsInput, (opts) => client.metaChanges(params, opts))`

And it will automatically handle tapAndCatch. 

Same can be done for promise as well:

`errorReporter.handlePromise(UserAction.Explicit, optsInput, ({skipIfImplicit, opts}) => client.metaChanges(params, opts))`

I think this will simply the 2 step process and won't have to handle finally manually everywhere.

Reply:
Yeah I thought about doing this and thought I was gonna be drinking too much typescript koolaid. But I'm feeling thirsty so will see if I can do something that feels nice like that before landing.

Reply:
Pushed something even smoother, check out https://github.com/sourcegraph/openctx/pull/142/commits/5124ce7a7341c1d8d94b4417493c118f43fec4dc